### ISSUE TITLE
Measure domain disparity

### ISSUE THREAD

**Issue by rpryzant** [rpryzant/japanese_corpus#13]
Since our whole schtick is combining datasets from different domains, we should provide some concrete measure of how different those domains are. 

[[Domain-Adversarial Training of Neural Networks]](https://arxiv.org/abs/1505.07818) gives a heuristic for doing just this. The heuristic is called **Proxy A-distance** (PAD). The algorithm is:

- Make a new dataset with a bunch of examples from domain D1 and domain D2. 
- Label examples from D1 with 0, examples from D2 with 1.
- Train a classifier on a subset of U
- Get error of this classifier on a separate subset of U. Call this error `e`. 
- set `PAD = 2 ( 1 - 2 e )` as your distance measure. 

The bigger the `e`, the smaller the PAD (since e is in [0, 1]). So PAD is basically saying that the harder two domains are to differentiate, the smaller the difference is between them. It is a measure of similarity, so the smaller the PAD, the closer the domains.

**Comment by rpryzant** [comment]
I can write a quick RNN classifier to do this.

**Comment by dennybritz** [comment]
Sounds like a great idea that will add to the paper. The other way I thought about measuring domain disparity (indirectly) was to train an NMT model on D1 and eval on D2, plus the other way around, and compare that to the number you get from training on D1 and eval on D1, etc.

**Comment by rpryzant** [comment]
I think that would work too, but it might be better to stick to existing precedent for now. There's some theory behind PAD that we can cite to make things look more official.

**Comment by rpryzant** [comment]
This is going to live in a standalone repo (https://github.com/rpryzant/mt_proxy_distance).

I couldn't get LSTMs to work very well (not sure why), so I concat the learned word vectors for a sequence and pass that through some fc layers. I'm also going to do a bag-of-words SVM. I didn't push too hard on the LSTM because I don't think the model architecture is very important. We just want to get some notion of a reasonable classifier's error.

**Comment by rpryzant** [comment]
This is done, apart from documentation/refactor which I will do later today.

I tried a few classifiers, and ended up only keeping a bag-of-words SVM. This is because
* I want to make this an easy-to-use tool for other people, they shouldn't be limited to CUDA-enabled machines to calculate PAD
* PAD values are by themselves meaningless. They really only make sense in a comparative setting. All we care about is what kind of error a reasonable classifier would get on the data, and that this classifier be consistent across dataset pairs. So I don't think there's any huge advantages to using really powerful deep classifiers. 
* SVM/liblinear solvers are wicket fast
